src/server/worker/pipeline/transform/registry.go

```diff
@@ -331,111 +377,69 @@ func (reg *registry) processDatums(ctx context.Context, pj *pendingJob, taskDoer
-		inputChan := make(chan *types.Any)
-		eg.Go(func() error {
-			defer close(inputChan)
-			commit := client.NewRepo(client.FileSetsRepoName).NewCommit("", fileSetID)
-			r, err := pachClient.GetFileTAR(commit, "/*")
+		return task.DoBatch(pachClient.Ctx(), taskDoer, inputs, func(_ int64, output *types.Any, err error) error {
 			if err != nil {
 				return err
 			}
-			if err := tarutil.Iterate(r, func(f tarutil.File) error {
-				buf := &bytes.Buffer{}
-				if err := f.Content(buf); err != nil {
-					return errors.EnsureStack(err)
-				}
-				input := &types.Any{}
-				if err := proto.Unmarshal(buf.Bytes(), input); err != nil {
-					return errors.EnsureStack(err)
-				}
-				select {
-				case inputChan <- input:
-				case <-ctx.Done():
-					return errors.EnsureStack(ctx.Err())
-				}
-				return nil
-			}, true); err != nil && !pfsserver.IsFileNotFoundErr(err) {
+			result, err := deserializeCreateDatumSetsTaskResult(output)
+			if err != nil {
 				return err
 			}
+			datumSets = append(datumSets, result.DatumSets...)
 			return nil
 		})
-		eg.Go(func() error {
-			return errors.EnsureStack(taskDoer.Do(
-				ctx,
-				inputChan,
-				func(_ int64, output *types.Any, err error) error {
-					if err != nil {
-						return err
-					}
-					data, err := deserializeDatumSet(output)
-					if err != nil {
-						return err
-					}
-					if _, err := pachClient.PfsAPIClient.AddFileSet(
-						pachClient.Ctx(),
-						&pfs.AddFileSetRequest{
-							Commit:    pj.commitInfo.Commit,
-							FileSetId: data.OutputFileSetId,
-						},
-					); err != nil {
-						return grpcutil.ScrubGRPC(err)
-					}
-					if _, err := pachClient.PfsAPIClient.AddFileSet(
-						pachClient.Ctx(),
-						&pfs.AddFileSetRequest{
-							Commit:    pj.metaCommitInfo.Commit,
-							FileSetId: data.MetaFileSetId,
-						},
-					); err != nil {
-						return grpcutil.ScrubGRPC(err)
-					}
-					if err := datum.MergeStats(stats, data.Stats); err != nil {
-						return err
-					}
-					pj.saveJobStats(data.Stats)
-					return pj.writeJobInfo()
-				},
-			))
-		})
-		return errors.EnsureStack(eg.Wait())
-	}))
+	}); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return datumSets, nil
+}
+
+func createSetSpec(pj *pendingJob) (*datum.SetSpec, error) {
+	var setSpec *datum.SetSpec
+	datumSetsPerWorker := defaultDatumSetsPerWorker
+	if pj.driver.PipelineInfo().Details.DatumSetSpec != nil {
+		setSpec = &datum.SetSpec{
+			Number:    pj.driver.PipelineInfo().Details.DatumSetSpec.Number,
+			SizeBytes: pj.driver.PipelineInfo().Details.DatumSetSpec.SizeBytes,
+		}
+		datumSetsPerWorker = pj.driver.PipelineInfo().Details.DatumSetSpec.PerWorker
+	}
+	// When the datum set spec is not set, evenly distribute the datums.
+	if setSpec == nil || (setSpec.Number == 0 && setSpec.SizeBytes == 0) {
+		concurrency, err := pj.driver.ExpectedNumWorkers()
+		if err != nil {
+			return nil, errors.EnsureStack(err)
+		}
+		numDatums := pj.ji.DataTotal - pj.ji.DataSkipped
```

Comment:
I'm surprised you caught this. I didn't want to try and account for that since we are planning to remove parallel jobs shortly after this. This isn't a correctness concern, it just means we may underutilize workers in certain cases. I don't think we will have a release before parallel jobs is gone, and even if we do it will be an alpha. I'm glad you reviewed thoroughly enough to catch this though.